Mathlib/Geometry/Manifold/ContMDiff.lean

```diff
@@ -2030,6 +2030,83 @@ nonrec theorem Smooth.smul {f : M → 𝕜} {g : M → V} (hf : Smooth I 𝓘(
   hf.smul hg
 #align smooth.smul Smooth.smul
 
+/-! ### Smoothness of an open embedding and its inverse -/
+
+section
+
+variable (I)
+  [Nonempty M] {e : M → H} (h : OpenEmbedding e)
+  [Nonempty M'] {e' : M' → H'} (h' : OpenEmbedding e')
+  {n : WithTop ℕ}
+
+lemma contMDiff_openEmbedding :
+    @ContMDiff _ _ _ _ _ _ _ I _ _ h.singletonChartedSpace _ _ _ _ _ I _ _ _ n e := by
+  haveI := h.singleton_smoothManifoldWithCorners I
+  rw [@contMDiff_iff _ _ _ _ _ _ _ _ _ _ h.singletonChartedSpace]
+  use h.continuous
+  intros x y -- show the function is actually the identity on the range of I ∘ e
+  apply contDiffOn_id.congr
+  intros z hz -- factorise into the chart (=e) and the model (=id)
+  rw [extChartAt_coe, @extChartAt_coe_symm _ _ _ _ _ _ _ _ _ _ _ h.singletonChartedSpace,
+    chartAt_self_eq]
+  repeat rw [Function.comp_apply]
+  rw [LocalHomeomorph.refl_apply, id.def, LocalHomeomorph.singletonChartedSpace_chartAt_eq,
+    h.toLocalHomeomorph_right_inv]
+  · rw [id.def, I.right_inv]
+    apply Set.mem_of_subset_of_mem _ hz.1
+    exact @extChartAt_target_subset_range _ _ _ _ _ _ _ _ _ I x h.singletonChartedSpace
+  · rw [ModelWithCorners.symm] -- show hz implies z is in range of I ∘ e
+    have := hz.1
+    rw [@extChartAt_target _ _ _ _ _ _ _ _ _ _ h.singletonChartedSpace] at this
+    have := this.1
+    rw [Set.mem_preimage, LocalHomeomorph.singletonChartedSpace_chartAt_eq,
+      h.toLocalHomeomorph_target] at this
+    exact this
+
+variable {I}
+lemma contMDiffOn_openEmbedding_symm :
+    @ContMDiffOn _ _ _ _ _ _ _ I _ _ _ _ _ _ _ _ I _ _ h.singletonChartedSpace
+      n (OpenEmbedding.toLocalHomeomorph e h).symm (Set.range e) := by
+  haveI := h.singleton_smoothManifoldWithCorners I
+  rw [@contMDiffOn_iff _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ h.singletonChartedSpace]
+  constructor
+  · rw [←h.toLocalHomeomorph_target]
+    apply (h.toLocalHomeomorph e).continuousOn_symm
+  · intros z hz -- show the function is actually the identity on the range of I ∘ e
+    apply contDiffOn_id.congr
+    intros z hz -- factorise into the chart (=e) and the model (=id)
+    rw [@extChartAt_coe _ _ _ _ _ _ _ _ _ _ _ h.singletonChartedSpace,
+      extChartAt_coe_symm, chartAt_self_eq]
+    simp_rw [Function.comp_apply]
+    rw [LocalHomeomorph.refl_symm, LocalHomeomorph.refl_apply, id.def,
+      LocalHomeomorph.singletonChartedSpace_chartAt_eq, LocalHomeomorph.right_inv]
+    · rw [ModelWithCorners.right_inv]
+      · rfl
+      apply Set.mem_of_subset_of_mem _ hz.1
+      exact extChartAt_target_subset_range _ _ --show hz implies z is in range of I ∘ e
+    · rw [h.toLocalHomeomorph_target, ModelWithCorners.symm, ←Set.mem_preimage]
+      have := hz.2
+      rw [Set.preimage_inter] at this
```

Comment:
`open`ing Set might also be useful here (I count 16 overall occurences in this file).

Reply:
Turns out it's already opened. I've removed all 16 `Set.`